### ISSUE TITLE
MoM architectural decisions

### ISSUE THREAD

**Issue by buep** [Praqma/git-phlow#234]
*This issue can be closed when the following architectural decisions have been documented with the code or documentation.* 

## We are only compliant with the Jira default workflow 

<img width="779" alt="jira workflow" src="https://user-images.githubusercontent.com/1579826/30159329-832ef7f2-93c8-11e7-9650-8fa4cc99e39c.png">

Image from https://confluence.atlassian.com/adminjiraserver071/working-with-workflows-802592661.html matching Jira server 7.1

The Jira support SoW states

> The solution will require you to be compliant in Jira with the Pragmatic workflow which means that an Jira issues should be able to have the Jira workflow statuses ‘IN PROGRESS’ and ‘ CLOSED’.

It was slightly formulated in Jira terms wrong, because "statuses" should have been status categories, because that will work better and make it compliant with even more workflows than the default.

In the image you see transition names on the arrows and status names in the boxes. Their color on the other hand defines the status category. One status category is called `in progress`.

By using status category we achieve to be compliant with many more workflows as status or transition doesn't have to have a special name, as long and the category is correct.

If there are more statuses from current status to a new status with statuscategory `in progress` we take the first one in the returned list in Jira.

## Work-on status change failure still establish a work space

Jira, Github and other issue systems will all follow this workflow with work-on

* set assignee
* set status (on github it is a lable, in jira a transition will happen)
* establish git workspace

(Side note: this generic workflow or any issues system must be implemented only once in the code, but call by delegation or like the correct issues system implementation - this is important, so we always ensure all issues systems follow the same workflow)

On github these operations will always work - nothing on the issues limits this. In Jira, and possible other systems, there can be limitations. In Jira there can be Conditions and Validator that block the transition. For example issue needs to have an estimate before being put in progress.

We have chose these are out of our control in git phlow, and it will be to complex to implement some kind of dry-run with trial transition and possible roll-back.

Further both set assignee and set status is independent operations in both Jira and Github and can't be done atomic anyway.

Thus each of the steps are done one by one, errors are reported as warnings to the user and finally the workspace is established.
Should the workspace establishment fail - it is considered a real error though, not just a warning.

Should an warning be issue, user have to sort out the problem using the issue system UI for example. The messages to the user should thus explain what was tried and what failed.

## Work-on do not check issue status

You can write `work-on #anyissues` - we will not check if it is closed or any status, but it will be change status as usual (if possible).
Basically it will allow you to work-on an closed issues. 
* In github it will not be re-opened
* In jira is will be re-opened if there is a status of status category `in progress` from the state smart commit puts the issue in when the word "Close" is detected. In the default workflow this is not possible.

_ So this means we haven't implemented the exact same behavior in Github and Jira, because in Github we do not try to re-open the issues, while we in Jira try to but most likely fail._

So the end result is workon and closed issues will assign yourself to the issue and give you a workspace.
